File: app/bot/handlers/private.py
from __future__ import annotations

from aiogram import F, Router
from aiogram.enums import ChatType, ParseMode
from aiogram.exceptions import TelegramBadRequest
from aiogram.filters import Command
from aiogram.types import InputMediaPhoto, InputMediaVideo, Message

from app.bot.ui import DISABLED_LINK_PREVIEW, append_original_link
from app.formatters.telegram import CAPTION_LIMIT, MESSAGE_LIMIT
from app.providers.base import TweetMedia
from app.services import AccessService, ShareResult, TweetShareService
from app.utils.urls import extract_first_tweet_url

router = Router(name="private")
router.message.filter(F.chat.type == ChatType.PRIVATE)

INVALID_LINK_TEXT = "🔗 Пришлите ссылку на пост X/Twitter, например https://x.com/user/status/123"
FETCH_ERROR_TEXT = (
    "⚠️ Не удалось получить пост. Возможно, он удален, приватный или временно недоступен."
)


@router.message(Command("start"))
async def start(message: Message, access_service: AccessService) -> None:
    assert message.from_user is not None
    has_access = await access_service.has_access(message.from_user.id)
    status = "🟢 доступ открыт" if has_access else "🔒 доступ закрыт"
    await message.answer(
        "👋 <b>Xtract Bot</b> помогает красиво пересылать посты X/Twitter в Telegram.\n\n"
        f"🆔 Ваш Telegram ID: <code>{message.from_user.id}</code>\n"
        f"📌 Статус: {status}\n\n"
        "📨 Отправьте ссылку на пост после получения доступа.",
        parse_mode=ParseMode.HTML,
        link_preview_options=DISABLED_LINK_PREVIEW,
    )


@router.message(Command("help"))
async def help_command(message: Message) -> None:
    await message.answer(
        "📖 <b>Как пользоваться ботом</b>\n\n"
        "📨 Отправьте ссылку на пост X/Twitter в личный чат с ботом.\n"
        "✅ Поддерживаются: x.com, twitter.com, mobile.twitter.com, vxtwitter.com.\n\n"
        "🔍 <b>Inline режим:</b> введите "
        "<code>@bot_username &lt;ссылка&gt;</code> в любом чате.\n\n"
        "🆔 /id — покажет ваш Telegram ID.",
        parse_mode=ParseMode.HTML,
        link_preview_options=DISABLED_LINK_PREVIEW,
    )


@router.message(Command("id"))
async def id_command(message: Message) -> None:
    assert message.from_user is not None
    await message.answer(
        f"🆔 Ваш Telegram ID: <code>{message.from_user.id}</code>", parse_mode=ParseMode.HTML
    )


@router.message(F.text)
async def handle_text(message: Message, tweet_share_service: TweetShareService) -> None:
    assert message.from_user is not None
    text = message.text or ""
    if text.startswith("/"):
        await message.answer("❓ Неизвестная команда. Используйте /help.")
        return
    if extract_first_tweet_url(text) is None:
        await message.answer(INVALID_LINK_TEXT, link_preview_options=DISABLED_LINK_PREVIEW)
        return

    result = await tweet_share_service.process_text(
        text,
        telegram_user_id=message.from_user.id,
        chat_id=message.chat.id,
        mode="private",
    )
    await _send_share_result(message, result)


async def _send_share_result(message: Message, result: ShareResult) -> None:
    if result.status == "invalid_url":
        await message.answer(INVALID_LINK_TEXT, link_preview_options=DISABLED_LINK_PREVIEW)
        return
    if not result.ok or result.post is None:
        await message.answer(FETCH_ERROR_TEXT)
        return

    post = result.post
    original_url = _result_original_url(result)
    caption = append_original_link(post.caption_html, original_url, limit=CAPTION_LIMIT)
    text = append_original_link(post.html, original_url, limit=MESSAGE_LIMIT)

    if post.media:
        await _send_media(message, list(post.media), caption=caption, fallback_text=text)
        return

    await message.answer(
        text,
        parse_mode=ParseMode.HTML,
        link_preview_options=DISABLED_LINK_PREVIEW,
    )


async def _send_media(
    message: Message,
    media: list[TweetMedia],
    *,
    caption: str,
    fallback_text: str,
) -> None:
    if len(media) == 1:
        await _send_one(message, media[0], caption=caption, fallback_text=fallback_text)
        return

    try:
        await message.answer_media_group(
            [
                _input_group_media(item, caption if index == 0 else None)
                for index, item in enumerate(media)
            ]
        )
        return
    except TelegramBadRequest:
        pass

    preview_group = _preview_input_group(media, caption)
    if preview_group is not None:
        try:
            await message.answer_media_group(preview_group)
            return
        except TelegramBadRequest:
            pass

    await message.answer(
        fallback_text,
        parse_mode=ParseMode.HTML,
        link_preview_options=DISABLED_LINK_PREVIEW,
    )


async def _send_one(
    message: Message,
    item: TweetMedia,
    *,
    caption: str,
    fallback_text: str,
) -> None:
    try:
        await _send_single_media(message, item, caption=caption)
        return
    except TelegramBadRequest:
        pass
    if item.preview_url:
        try:
            await message.answer_photo(
                item.preview_url,
                caption=caption,
                parse_mode=ParseMode.HTML,
                show_caption_above_media=True,
            )
            return
        except TelegramBadRequest:
            pass
    await message.answer(
        fallback_text,
        parse_mode=ParseMode.HTML,
        link_preview_options=DISABLED_LINK_PREVIEW,
    )


async def _send_single_media(
    message: Message,
    item: TweetMedia,
    *,
    caption: str | None,
) -> None:
    parse_mode = ParseMode.HTML if caption else None
    show_above = True if caption else None
    if item.type == "photo":
        await message.answer_photo(
            item.url,
            caption=caption,
            parse_mode=parse_mode,
            show_caption_above_media=show_above,
        )
    elif item.type == "gif":
        await message.answer_animation(
            item.url,
            caption=caption,
            parse_mode=parse_mode,
            show_caption_above_media=show_above,
            width=item.width,
            height=item.height,
            duration=_duration_seconds(item.duration_ms),
        )
    else:
        await message.answer_video(
            item.url,
            caption=caption,
            parse_mode=parse_mode,
            show_caption_above_media=show_above,
            width=item.width,
            height=item.height,
            duration=_duration_seconds(item.duration_ms),
        )


def _input_group_media(item: TweetMedia, caption: str | None):
    parse_mode = ParseMode.HTML if caption else None
    show_above = True if caption else None
    if item.type == "photo":
        return InputMediaPhoto(
            media=item.url,
            caption=caption,
            parse_mode=parse_mode,
            show_caption_above_media=show_above,
        )
    return InputMediaVideo(
        media=item.url,
        caption=caption,
        parse_mode=parse_mode,
        show_caption_above_media=show_above,
        width=item.width,
        height=item.height,
        duration=_duration_seconds(item.duration_ms),
    )


def _preview_input_group(media: list[TweetMedia], caption: str) -> list[InputMediaPhoto] | None:
    items: list[InputMediaPhoto] = []
    for item in media:
        url = item.url if item.type == "photo" else item.preview_url
        if not url:
            return None
        is_first = not items
        items.append(
            InputMediaPhoto(
                media=url,
                caption=caption if is_first else None,
                parse_mode=ParseMode.HTML if is_first else None,
                show_caption_above_media=True if is_first else None,
            )
        )
    return items if len(items) >= 2 else None


def _duration_seconds(duration_ms: int | None) -> int | None:
    if duration_ms is None:
        return None
    return max(1, round(duration_ms / 1000))


def _result_original_url(result: ShareResult) -> str:
    if result.tweet is not None:
        return result.tweet.url
    return result.normalized_url or result.source_url or "https://x.com"

File: src/bot/handlers/private.ts
import { Composer, GrammyError, InputFile, type InlineKeyboard } from "grammy";
import type { InputMediaPhoto, InputMediaVideo } from "grammy/types";

import { DISABLED_LINK_PREVIEW, openProfileButton, originalPostButton } from "@/bot/ui";
import type { AppContext } from "@/bot/context";
import { buildRichMessage, withAvatarEmoji } from "@/formatters/richMessage";
import type { TelegramPost } from "@/formatters/telegram";
import type { TweetMedia } from "@/providers/base";
import type { ProfileShareResult } from "@/services/profileShare";
import type { ShareResult } from "@/services/tweetShare";
import { extractFirstProfileUrl, extractFirstTweetUrl } from "@/utils/urls";

const INVALID_LINK_TEXT =
  "🔗 Пришлите ссылку на пост или профиль X/Twitter, например " +
  "https://x.com/user/status/123 или https://x.com/user";
const FETCH_ERROR_TEXT =
  "⚠️ Не удалось получить пост. Возможно, он удален, приватный или временно недоступен.";
const PROFILE_ERROR_TEXT =
  "⚠️ Не удалось получить профиль. Возможно, он удален, приватный или временно недоступен.";
// Caption shown inside the animated "generating" placeholder (rich draft).
const THINKING_CAPTION = "Загружаю пост…";

export const privateComposer = new Composer<AppContext>();

const privateChat = privateComposer.filter((ctx) => ctx.chat?.type === "private");

privateChat.command("start", async (ctx) => {
  await ctx.reply(
    "👋 <b>Xtract Bot</b> помогает красиво пересылать посты и профили X/Twitter в Telegram.\n\n" +
      "📨 Отправьте ссылку на пост X/Twitter в личный чат — бот вытащит текст, медиа " +
      "и оформит сообщение.\n" +
      "👤 Отправьте ссылку на профиль (например <code>https://x.com/user</code>) — бот " +
      "красиво оформит карточку профиля.\n" +
      "🔍 Или используйте inline режим: <code>@bot_username &lt;ссылка&gt;</code> в любом чате.\n\n" +
      "ℹ️ Подробнее: /help",
    { parse_mode: "HTML", link_preview_options: DISABLED_LINK_PREVIEW },
  );
});

privateChat.command("help", async (ctx) => {
  const isAdmin = ctx.from ? ctx.services.access.isAdmin(ctx.from.id) : false;
  const baseHelp =
    "📖 <b>Как пользоваться ботом</b>\n\n" +
    "📨 Отправьте ссылку на пост X/Twitter в личный чат с ботом.\n" +
    "👤 Отправьте ссылку на профиль (<code>https://x.com/user</code>) — бот оформит " +
    "карточку профиля.\n" +
    "✅ Поддерживаются: x.com, twitter.com, mobile.twitter.com, vxtwitter.com.\n\n" +
    "🔍 <b>Inline режим:</b> введите " +
    "<code>@bot_username &lt;ссылка&gt;</code> в любом чате.\n\n" +
    "🆔 /id — покажет ваш Telegram ID.";
  const adminHelp =
    "\n\n🛠 <b>Команды администратора</b>\n" +
    "/panel — сводка по боту и whitelist\n" +
    "/stats [telegram_id] — статистика (общая или по пользователю)\n" +
    "/users — список пользователей в whitelist\n" +
    "/allow &lt;telegram_id&gt; — добавить в whitelist\n" +
    "/deny &lt;telegram_id&gt; — удалить из whitelist\n" +
    "/whitelist [on|off] — статус/включить/выключить whitelist\n" +
    "/translate [on|off] — перевод на русский в inline (beta)\n" +
    "/health — проверка БД и провайдера";
  await ctx.reply(baseHelp + (isAdmin ? adminHelp : ""), {
    parse_mode: "HTML",
    link_preview_options: DISABLED_LINK_PREVIEW,
  });
});

privateChat.command("id", async (ctx) => {
  if (!ctx.from) return;
  await ctx.reply(`🆔 Ваш Telegram ID: <code>${ctx.from.id}</code>`, { parse_mode: "HTML" });
});

privateChat.on("message:text", async (ctx) => {
  if (!ctx.from) return;
  const text = ctx.message.text;
  if (text.startsWith("/")) {
    await ctx.reply("❓ Неизвестная команда. Используйте /help.");
    return;
  }

  // A status link is a post; otherwise a bare handle link is a profile.
  const tweetUrl = extractFirstTweetUrl(text);
  const profileUrl = tweetUrl ? null : extractFirstProfileUrl(text);

  // Show an animated "generating" placeholder while we fetch (private chats only).
  if (tweetUrl !== null || profileUrl !== null) {
    await sendThinkingDraft(ctx);
  }

  if (profileUrl !== null) {
    const result = await ctx.services.profileShare.processUrl(profileUrl, {
      telegramUserId: ctx.from.id,
      chatId: ctx.chat.id,
      mode: "private",
    });
    await sendProfileResult(ctx, result);
    return;
  }

  const result = await ctx.services.tweetShare.processText(text, {
    telegramUserId: ctx.from.id,
    chatId: ctx.chat.id,
    mode: "private",
  });
  await sendShareResult(ctx, result);
});

// Stream a short-lived rich draft with a <tg-thinking> animation. It's a
// nice-to-have, so any failure is swallowed and the real reply still follows.
async function sendThinkingDraft(ctx: AppContext): Promise<void> {
  try {
    await ctx.replyWithRichMessageDraft(
      { html: `<tg-thinking>${THINKING_CAPTION}</tg-thinking>` },
      { draft_id: ctx.message?.message_id ?? 1 },
    );
  } catch (error) {
    if (!(error instanceof GrammyError)) throw error;
  }
}

async function sendShareResult(ctx: AppContext, result: ShareResult): Promise<void> {
  if (result.status === "invalid_url") {
    await ctx.reply(INVALID_LINK_TEXT, { link_preview_options: DISABLED_LINK_PREVIEW });
    return;
  }
  if (!result.ok || result.post === null) {
    await ctx.reply(FETCH_ERROR_TEXT);
    return;
  }

  const url = result.tweet?.url ?? result.normalizedUrl ?? "";
  const avatarEmoji = await resolveAvatarEmoji(ctx, result.tweet?.authorAvatarUrl);
  await replyWithPost(ctx, result.post, originalPostButton(url), avatarEmoji);
}

async function sendProfileResult(ctx: AppContext, result: ProfileShareResult): Promise<void> {
  if (result.status === "invalid_url") {
    await ctx.reply(INVALID_LINK_TEXT, { link_preview_options: DISABLED_LINK_PREVIEW });
    return;
  }
  if (!result.ok || result.post === null) {
    await ctx.reply(PROFILE_ERROR_TEXT);
    return;
  }

  const url = result.profile?.url ?? result.normalizedUrl ?? "";
  const avatarEmoji = await resolveAvatarEmoji(ctx, result.profile?.avatarUrl);
  await replyWithPost(ctx, result.post, openProfileButton(url), avatarEmoji);
}

interface AvatarEmoji {
  id: string;
  glyph: string;
}

// Best-effort: turn the avatar into a custom emoji to show inline. Returns
// undefined when the feature is off or anything fails (it never throws).
async function resolveAvatarEmoji(
  ctx: AppContext,
  avatarUrl: string | null | undefined,
): Promise<AvatarEmoji | undefined> {
  const service = ctx.services.avatarEmoji;
  if (!service || !avatarUrl) return undefined;
  const id = await service.resolve(avatarUrl);
  return id ? { id, glyph: service.fallbackGlyph } : undefined;
}

// Send an already-formatted post: prefer a Rich Message (long text, up to ~32k
// chars, plus an inline media carousel), then fall back to the legacy senders.
// When an avatar emoji is given, try it first and retry without it if Telegram
// rejects the custom emoji (e.g. the bot isn't eligible to send one).
async function replyWithPost(
  ctx: AppContext,
  post: TelegramPost,
  button: InlineKeyboard,
  avatarEmoji?: AvatarEmoji,
): Promise<void> {
  const rich = buildRichMessage(post);
  const candidates = avatarEmoji
    ? [withAvatarEmoji(rich, avatarEmoji.id, avatarEmoji.glyph), rich]
    : [rich];
  for (const candidate of candidates) {
    try {
      await ctx.replyWithRichMessage(candidate, { reply_markup: button });
      return;
    } catch (error) {
      if (!(error instanceof GrammyError)) throw error;

      console.error("failed to send rich message", error);
    }
  }

  if (post.media.length > 0) {
    await sendMedia(ctx, [...post.media], {
      caption: post.captionHtml,
      captionGroup: `${post.captionHtml}\n\n${post.linkHtml}`,
      fallbackText: post.html,
      replyMarkup: button,
    });
    return;
  }

  await ctx.reply(post.html, {
    parse_mode: "HTML",
    link_preview_options: DISABLED_LINK_PREVIEW,
    reply_markup: button,
  });
}

interface SendMediaOptions {
  caption: string;
  captionGroup: string;
  fallbackText: string;
  replyMarkup: InlineKeyboard;
}

async function sendMedia(
  ctx: AppContext,
  media: TweetMedia[],
  options: SendMediaOptions,
): Promise<void> {
  if (media.length >= 2) {
    try {
      await ctx.replyWithMediaGroup(
        media.map((item, index) =>
          inputGroupMedia(item, index === 0 ? options.captionGroup : null),
        ),
      );
      return;
    } catch (error) {
      if (!(error instanceof GrammyError)) throw error;
    }

    const previewGroup = previewInputGroup(media, options.captionGroup);
    if (previewGroup !== null) {
      try {
        await ctx.replyWithMediaGroup(previewGroup);
        return;
      } catch (error) {
        if (!(error instanceof GrammyError)) throw error;
      }
    }
  }

  let captionSent = false;
  let anySent = false;
  for (const item of media) {
    const itemCaption = captionSent ? null : options.caption;
    const itemMarkup = captionSent ? null : options.replyMarkup;
    if (await trySendOne(ctx, item, itemCaption, itemMarkup)) {
      anySent = true;
      if (itemCaption !== null) captionSent = true;
    }
  }

  if (!anySent || !captionSent) {
    await ctx.reply(options.fallbackText, {
      parse_mode: "HTML",
      link_preview_options: DISABLED_LINK_PREVIEW,
      reply_markup: options.replyMarkup,
    });
  }
}

async function trySendOne(
  ctx: AppContext,
  item: TweetMedia,
  caption: string | null,
  replyMarkup: InlineKeyboard | null,
): Promise<boolean> {
  try {
    await sendSingleMedia(ctx, item, caption, replyMarkup);
    return true;
  } catch (error) {
    if (!(error instanceof GrammyError)) throw error;
  }
  if (item.previewUrl) {
    try {
      await ctx.replyWithPhoto(item.previewUrl, {
        caption: caption ?? undefined,
        parse_mode: caption ? "HTML" : undefined,
        reply_markup: replyMarkup ?? undefined,
      });
      return true;
    } catch (error) {
      if (!(error instanceof GrammyError)) throw error;
    }
  }
  return false;
}

async function sendSingleMedia(
  ctx: AppContext,
  item: TweetMedia,
  caption: string | null,
  replyMarkup: InlineKeyboard | null,
): Promise<void> {
  const parseMode = caption ? ("HTML" as const) : undefined;
  if (item.type === "photo") {
    await ctx.replyWithPhoto(item.url, {
      caption: caption ?? undefined,
      parse_mode: parseMode,
      reply_markup: replyMarkup ?? undefined,
    });
    return;
  }
  if (item.type === "gif") {
    await ctx.replyWithAnimation(item.url, {
      caption: caption ?? undefined,
      parse_mode: parseMode,
      width: item.width ?? undefined,
      height: item.height ?? undefined,
      duration: durationSeconds(item.durationMs) ?? undefined,
      reply_markup: replyMarkup ?? undefined,
    });
    return;
  }
  await ctx.replyWithVideo(item.url, {
    caption: caption ?? undefined,
    parse_mode: parseMode,
    width: item.width ?? undefined,
    height: item.height ?? undefined,
    duration: durationSeconds(item.durationMs) ?? undefined,
    reply_markup: replyMarkup ?? undefined,
  });
}

function inputGroupMedia(
  item: TweetMedia,
  caption: string | null,
): InputMediaPhoto | InputMediaVideo {
  if (item.type === "photo") {
    return {
      type: "photo",
      media: item.url,
      caption: caption ?? undefined,
      parse_mode: caption ? "HTML" : undefined,
    };
  }
  return {
    type: "video",
    media: item.url,
    caption: caption ?? undefined,
    parse_mode: caption ? "HTML" : undefined,
    width: item.width ?? undefined,
    height: item.height ?? undefined,
    duration: durationSeconds(item.durationMs) ?? undefined,
  };
}

function previewInputGroup(media: TweetMedia[], caption: string): InputMediaPhoto[] | null {
  const items: InputMediaPhoto[] = [];
  for (const item of media) {
    let url: string;
    if (item.type === "photo") {
      url = item.previewUrl ?? item.url;
    } else if (item.previewUrl) {
      url = item.previewUrl;
    } else {
      return null;
    }
    const isFirst = items.length === 0;
    items.push({
      type: "photo",
      media: url,
      caption: isFirst ? caption : undefined,
      parse_mode: isFirst ? "HTML" : undefined,
    });
  }
  return items.length >= 2 ? items : null;
}

function durationSeconds(durationMs: number | null): number | null {
  if (durationMs === null) return null;
  return Math.max(1, Math.round(durationMs / 1000));
}

// Re-export to satisfy bundlers in case helpers tree-shake unevenly.
export { InputFile };
